# Sowing Success: How Machine Learning Helps Farmers Select the Best Crops

![Farmer in a field](farmer_in_a_field.jpg)

**Introduction:**

Assessing soil condition is crucial for farmers to determine the optimal crop for their fields. However, traditional methods of measuring soil metrics like nitrogen, phosphorous, potassium levels, and pH value are expensive and time-consuming. As a result, farmers often face budget constraints and must prioritize which metrics to measure.

In this project, the objective is to leverage machine learning techniques to assist farmers in selecting the best crop for their fields based on soil condition. By analyzing a dataset containing soil metrics and corresponding crop types, we aim to build multi-class classification models to predict the type of crop that would thrive in a given soil environment. Additionally, we will identify the single most important feature for predictive performance, aiding farmers in making informed decisions to maximize crop yield.

**Dataset Description:**

The dataset, named "soil_measures.csv," includes the following features:

- "N": Nitrogen content ratio in the soil
- "P": Phosphorous content ratio in the soil
- "K": Potassium content ratio in the soil
- "pH": pH value of the soil
- "crop": Categorical values representing various crop types (target variable)

Each row in the dataset represents measurements of soil metrics in a particular field, along with the corresponding optimal crop specified in the "crop" column.

**Project Goals:**

1. Build multi-class classification models to predict the type of crop based on soil metrics.
2. Identify the single most important feature for predictive performance.
3. Assist farmers in selecting the best crop for their fields by providing actionable insights derived from machine learning analysis.

Through this project, we aim to demonstrate how machine learning can be utilized to optimize crop selection and contribute to the success of agricultural practices, ultimately benefiting farmers and enhancing agricultural productivity.

In [2]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics

# Load the dataset
crops = pd.read_csv("soil_measures.csv")

# Split the dataset into features (X) and target variable (y)
X = crops.drop(columns=['crop'])
y = crops['crop']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train a logistic regression model
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

# Evaluate the model
accuracy = model.score(X_test, y_test)
print("Accuracy:", accuracy)

# Get feature importance
feature_importance = model.coef_[0]
best_feature_index = feature_importance.argmax()
best_feature_name = X.columns[best_feature_index]

# Create a dictionary containing the best predictive feature name and its evaluation score
best_predictive_feature = {best_feature_name: accuracy}
print("Best predictive feature:", best_predictive_feature)


Accuracy: 0.6113636363636363
Best predictive feature: {'K': 0.6113636363636363}


# Conclusion

After analyzing the soil metrics dataset and applying machine learning techniques, we have achieved an accuracy of 61.14% in predicting the type of crop based on soil condition. The best predictive feature identified for this task is the Potassium content ratio ("K") in the soil, which yielded an accuracy of 61.14%.

This project demonstrates the potential of machine learning in assisting farmers with crop selection by leveraging soil metrics data. By accurately predicting the optimal crop type based on soil condition, farmers can make informed decisions to maximize crop yield and optimize agricultural productivity.

Moving forward, further refinement of the classification models and incorporation of additional features could potentially enhance predictive performance. Additionally, integrating real-time soil data collection and analysis through IoT devices could enable more dynamic and precise crop recommendations, ultimately contributing to sustainable and efficient agricultural practices.
